static-build/posts/2021/09/cors/index.md

```diff
@@ -0,0 +1,354 @@
+  <allow-access-from domain="www.example.com" />
+  <allow-http-request-headers-from domain="*.adobe.com" headers="SOAPAction" />
+</cross-domain-policy>
+```
+
+There are a few issues with this:
+
+- It changes the behaviour for the whole origin. You can imagine a similar format that lets you specify rules for particular resources, but the resource would start to get quite large.
+- You end up with two request, one for the `/crossdomain.xml`, and one for the actual resource. This becomes more of an issue the bigger `/crossdomain.xml` gets.
+- For larger sites built by multiple teams, you end up with issues over ownership of `/crossdomain.xml`.
+
+## In-resource opt-in
+
+To cut down the number of requests, the opt-in could be granted within the resource itself. This technique was proposed by the W3C Voice Browser Working Group back in 2005, using an XML processing instruction:
+
+```
+<?access-control allow="*.example.com" deny="*.visitors.example.com"?>
+```
+
+But what if the resource wasn't XML? Well, the opt-in would need to be in a different format.
+
+This is kinda where things landed for frame-to-frame communication. Both sides opt-in using [`postMessage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage), and can declare the origin they're happy to communicate with.
+
+But what about accessing the raw bytes of the resource? In that case it doesn't make sense to use resource-specific metadata for the opt-in. And besides, HTTP already has a place for resource metadata…
+
+## HTTP header opt-in
+
+The proposal by the Voice Browser Working Group was generalised using HTTP headers, and that became CORS.
+
+```
+Access-Control-Allow-Origin: *
+```
+
+# Making a CORS request
+
+Most modern web features require CORS by default, such as `fetch()`. The exception is modern features that are designed to support older features that don't use CORS, eg `<link rel="preload">`.
+
+Unfortunately there's no easy rule here. For example:
+
+```html
+<!-- Not a CORS request -->
+<script src="https://example.com/script.js"></script>
+<!-- CORS request -->
+<script type="module" src="https://example.com/script.js"></script>
+```
+
+The best way to figure it out is to try it and look at network DevTools. Cross-origin requests are sent with a [`Sec-Fetch-Mode` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Sec-Fetch-Mode) which will tell you if it's a CORS request or not.
+
+If an HTML element causes a no-CORS fetch, you can use the badly-named `crossorigin` attribute to switch it to a CORS request.
+
+```html
+<img crossorigin src="…" />
+<script crossorigin src="…"></script>
+<link crossorigin rel="stylesheet" href="…" />
+<link crossorigin rel="preload" as="font" href="…" />
+```
+
+When you switch these over to CORS, you get more visibility into the cross-origin resource:
+
+- You can paint the `<img>` to a `<canvas>` and read back the pixels.
+- You get more detailed stack traces for script.
```

Comment:
...in `window.onerror` -- this could be understood to apply to devtools

Reply:
It's not just in onerror is it? You'd get the details from a `catch` too right?

Reply:
I thought `catch` didn't mute errors, since it needs to be the same script that has the error that handles the error, so it can't leak information across origins. But I could be wrong.

https://html.spec.whatwg.org/multipage/webappapis.html#runtime-script-errors is for `window.onerror`

https://html.spec.whatwg.org/multipage/webappapis.html#the-hostpromiserejectiontracker-implementation is for `rejectionhandled` events

Reply:
> since it needs to be the same script that has the error that handles the error,

Uh, no, scratch that!

Reply:
I'll give it a test. I assumed if the throwing function was from a no-cors script, although it's caught in a same-origin script, it'd hide details of the stack

Reply:
http://software.hixie.ch/utilities/js/live-dom-viewer/saved/9659